contract/src/utils.rs

```diff
@@ -390,3 +365,336 @@ pub(crate) fn is_token_burned(token_identifier: &TokenIdentifier) -> bool {
+        NFTCoreError::InvalidMaximumNumberOfPages,
+    );
+
+    if back_fill_token_tracker >= tokens_minted_at_upgrade {
+        runtime::revert(NFTCoreError::InvalidNumberOfMintedTokens)
+    }
+
+    for token_identifier in owned_tokens_list.into_iter() {
+        let current_page_number = back_fill_token_tracker / PAGE_SIZE;
+        let current_page_index = back_fill_token_tracker % PAGE_SIZE;
+        let mut page_record = match storage::dictionary_get::<Vec<bool>>(
+            all_token_owners_uref,
+            &token_owner_item_key,
+        )
+        .unwrap_or_revert()
+        {
+            Some(page_record) => page_record,
+            None => vec![false; page_record_width as usize],
+        };
+        let _ = core::mem::replace(&mut page_record[current_page_number as usize], true);
+        storage::dictionary_put(all_token_owners_uref, &token_owner_item_key, page_record);
+        let single_page_uref = get_uref(
+            &format!("{}-{}", PAGE_DICTIONARY_PREFIX, current_page_number),
+            NFTCoreError::MissingStorageUref,
+            NFTCoreError::InvalidStorageUref,
+        );
+        let single_page_item_key = make_page_dictionary_item_key(&token_owner, current_page_number);
+        let mut single_page =
+            match storage::dictionary_get::<Vec<bool>>(single_page_uref, &single_page_item_key)
+                .unwrap_or_revert()
+            {
+                Some(single_page) => single_page,
+                None => vec![false; PAGE_SIZE as usize],
+            };
+        let _ = core::mem::replace(&mut single_page[current_page_index as usize], true);
+        storage::dictionary_put(single_page_uref, &single_page_item_key, single_page);
+        update_forward_and_reverse_token_trackers(back_fill_token_tracker, token_identifier);
+        back_fill_token_tracker += 1;
+    }
+
+    let back_fill_uref = get_uref(
+        BACKFILLED_TOKEN_TRACKER,
+        NFTCoreError::MissingBackFillTracker,
+        NFTCoreError::InvalidBackFillTracker,
+    );
+
+    storage::write(back_fill_uref, back_fill_token_tracker);
+}
+
+// This function is incredibly gas expensive
+// DO not use this function unless absolutely necessary.
+pub(crate) fn get_all_owned_token_identifiers(token_owner: Key) -> Vec<TokenIdentifier> {
+    let mut token_identifiers: Vec<TokenIdentifier> = vec![];
+    let token_item_key = get_owned_tokens_dictionary_item_key(token_owner);
+    let page_record = get_dictionary_value_from_key::<Vec<bool>>(ALL_TOKEN_OWNERS, &token_item_key)
+        .unwrap_or_revert_with(NFTCoreError::InvalidTokenOwner);
+    let identifier_mode: NFTIdentifierMode = get_stored_value_with_user_errors::<u8>(
+        IDENTIFIER_MODE,
+        NFTCoreError::MissingIdentifierMode,
+        NFTCoreError::InvalidIdentifierMode,
+    )
+    .try_into()
+    .unwrap_or_revert();
+    for (page_number, is_page_allocated) in page_record.into_iter().enumerate() {
+        if is_page_allocated {
```

Comment:
flip around to continue if not allocated

Reply:
Flipped as requested